File: AMF5/src/test/scala/scalaPlatform/StreamedRenderTest.scala
package scalaPlatform

import amf.apicontract.client.scala.OASConfiguration
import amf.core.client.scala.model.document.Document
import org.junit.runner.RunWith
import org.scalatest.flatspec.AsyncFlatSpec
import org.scalatest.matchers.must.Matchers.convertToAnyMustWrapper
import org.scalatest.matchers.should
import org.scalatestplus.junit.JUnitRunner
import org.yaml.builder.JsonOutputBuilder

import java.io.{File, FileOutputStream, FileWriter, OutputStreamWriter}
import java.util.Date
import scala.io.Source

@RunWith(classOf[JUnitRunner])
class StreamedRenderTest extends AsyncFlatSpec with should.Matchers with FileReader {

  "AMF" should "stream the rendering of Json-LD" in {
    val client = OASConfiguration.OAS20().baseUnitClient()
    val tmpFile = File.createTempFile("banking-api.json", new Date().getTime.toString)
    client.parse("file://src/test/resources/examples/banking-api.json") map { result =>
      result.baseUnit mustBe a[Document]
      val writer = new OutputStreamWriter(new FileOutputStream(tmpFile))
      val builder = JsonOutputBuilder(writer, prettyPrint = true)
      client.renderGraphToBuilder(result.baseUnit, builder)
      writer.flush()
      writer.close()
      val goldenContent = readResource("/examples/banking-api.flattened.jsonld")
      val writtenContent = using(Source.fromFile(tmpFile)) { source => source.mkString }
      goldenContent shouldEqual writtenContent
    }
  }
}

File: AMF5/src/test/java/javaPlatform/StreamedRenderTest.java
package javaPlatform;

import amf.apicontract.client.platform.AMFBaseUnitClient;
import amf.apicontract.client.platform.AMFDocumentResult;
import amf.apicontract.client.platform.OASConfiguration;
import org.junit.Test;
import org.yaml.builder.JsonOutputBuilder;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.nio.file.Files;
import java.util.Date;
import java.util.concurrent.ExecutionException;

import static org.junit.Assert.assertEquals;

public class StreamedRenderTest implements FileReader{

    @Test
    public void streamedTest() throws ExecutionException, InterruptedException, IOException {
        AMFBaseUnitClient client = OASConfiguration.OAS20().baseUnitClient();
        File tmpFile = File.createTempFile("banking-api.json", String.valueOf(new Date().getTime()));
        AMFDocumentResult result = client.parseDocument("file://src/test/resources/examples/banking-api.json").get();
        assert(result.conforms());
        Writer writer = new FileWriter(tmpFile);
        JsonOutputBuilder<Writer> jsonBuilder = JsonOutputBuilder.apply(writer, false);
        client.renderGraphToBuilder(result.baseUnit(), jsonBuilder);
        writer.flush();
        writer.close();
        String goldenContent = readResource("/examples/banking-api.flattened.jsonld");
        String writtenContent = String.join("", Files.readAllLines(tmpFile.toPath()));
        assertEquals(removeWhiteSpaces(writtenContent), removeWhiteSpaces(goldenContent));
    }

    private String removeWhiteSpaces(String input) {
        return input.replaceAll("\\s+", "");
    }
}
